src/java_bytecode/generate_java_generic_method.cpp

```diff
@@ -0,0 +1,296 @@
+    message_handler(message_handler)
+{}
+
+/// Generate a copy of a given generic method, specialized with
+/// the given concrete types and insert the method into the symbol
+/// table.
+/// \param generic_method The generic method to be specialized.
+/// \param concrete_types A map from generic type variables to concrete types
+/// with which to instantiate the generic method.
+/// \param symbol_table The symbol table into which the generated method will
+/// be inserted.
+/// \return The symbol that was inserting into the symbol_table, or the existing
+/// symbol if the method has already been specialized.
+const symbolt& generate_java_generic_methodt::operator()(
+  const symbolt &generic_method,
+  const type_variable_instantiationst &concrete_types,
+  symbol_tablet &symbol_table) const
+{
+  INVARIANT(generic_method.type.id()==ID_code, "Only code symbols");
+  INVARIANT(concrete_types.size()>0, "Should be at least one concrete type");
+  // Another invariant is that concrete_types.size() > the total number of
+  // type variables used in the method. However, determining that essentially
+  // means iterating through the whole method structure and duplicating a
+  // large chunk of the work we're about to do...
+
+
+  // We need to decorate the signature of the specialized method to
+  // differentiate the symbols of the generic method and the specialized
+  // method. There's a design decision to be made here - do we actually modify
+  // the java signature, or do we decorate it with additional
+  // qualifiers. We are taking the approach of adding decorations because
+  // that avoids the situation where the user may have already
+  // declared an overloaded method that might have the same signature
+  // as the specialized method we are about to create. By decorating the
+  // existing symbol, we ensure that no legal Java program could have already
+  // defined the symbol.
+  const std::string& signature_decoration=instantiation_decoration
+    (concrete_types);
+
+  // Check the method has not already been specialized
+  symbol_tablet::symbolst::const_iterator already_specialized=
+    symbol_table.symbols.find(
+      id2string(generic_method.name)+signature_decoration);
+  if(already_specialized!=symbol_table.symbols.end())
+    return already_specialized->second;
+
+  // Convert concrete types into a map for faster lookups
+  type_variable_instantiation_mapt concrete_type_map(concrete_types);
+
+  // Copy the generic method as a starting point for the specialization
+  symbolt specialized_method=generic_method;
+  code_typet &specialized_code=to_code_type(specialized_method.type);
+
+  // Handle a generic return type
+  typet &method_return_type=specialized_code.return_type();
+  if(is_java_generic_parameter(method_return_type))
+  {
+    instantiate_java_generic_parameter(
+      to_java_generic_parameter(method_return_type),
+      concrete_type_map);
+  }
+  else if(is_java_generic_type(method_return_type))
```

Comment:
And this is the case of `List<T>` for example. Consider adding a comment to make this explicit

Reply:
Good point - I'll add some comments to make this clearer.